java-checks/src/main/java/org/sonar/java/checks/security/AndroidPersistentUniqueIdentifierCheck.java

```diff
@@ -0,0 +1,125 @@
+  private static final String PHONE_NUMBER_MESSAGE = "Using a phone number puts user privacy at risk. Make sure it is safe here.";
+  private static final String ADVERTISING_ID_MESSAGE = "Using Advertising ID puts user privacy at risk. Make sure it is safe here.";
+  private static final String NON_RESETTABLE_PERSISTENT_ID_MESSAGE = "Using a non-resettable persistent identifier puts user privacy at risk. Make sure it is safe here.";
+
+  private static final MethodMatchers staticSettingsSecureGetStringFunMatcher =
+    MethodMatchers.create()
+      .ofTypes("android.provider.Settings$Secure")
+      .names("getString")
+      .addParametersMatcher("android.content.ContentResolver", "java.lang.String")
+      .build();
+
+  private static final Map<MethodMatchers, String> matchers = Map.of(
+    MethodMatchers.create()
+      .ofTypes("android.bluetooth.BluetoothAdapter")
+      .names("getAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.net.wifi.WifiInfo")
+      .names("getMacAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getSimSerialNumber", "getDeviceId", "getImei", "getMeid")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getLine1Number")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.SubscriptionManager")
+      .names("getPhoneNumber")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes(
+        "com.google.android.gms.ads.identifier.AdvertisingIdClient$Info",
+        "androidx.ads.identifier.AdvertisingIdInfo",
+        "com.huawei.hms.ads.identifier.AdvertisingIdClient$Info"
+      )
+      .names("getId")
+      .withAnyParameters()
+      .build(),
+    ADVERTISING_ID_MESSAGE,
+    staticSettingsSecureGetStringFunMatcher,
+    NON_RESETTABLE_PERSISTENT_ID_MESSAGE
+  );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.METHOD_REFERENCE);
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(matchers.keySet().toArray(MethodMatchers[]::new));
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    matchers.keySet().stream()
+      .filter(matcher -> matcher.matches(mit))
+      .findFirst()
+      .filter(matcher ->
+        matcher != staticSettingsSecureGetStringFunMatcher ||
+          mit.arguments().get(1).asConstant().map("android_id"::equals).orElse(false))
+      .ifPresent(matcher -> {
+        var tree = mit.methodSelect().is(Tree.Kind.MEMBER_SELECT)
```

Comment:
I couldn't cover the else branch of this ternary here, but I didn't want to just cast to `MemberSelectExpressionTree ` either. If you are confident that the cast will always succeed for a method invocation, I can remove the ternary. Otherwise I would keep it: it seems something that can easily change in the future, and being a bit defensive doesn't hurt.

Reply:
I think we can keep the uncovered branch because there is one (very very unlikely) case that can make the code crash without it.

If sonar-java analyzes the code of the file where the type and method we are trying to match is both defined and invoked/referenced, then we cannot rely on the method select being available.

For instance you can make the analyzer crash by adding the following segment to the body of `BluetoothAdapter` in `android/bluetooth/BluetoothAdapter.java` and removing the dead branch in the check
```java
  public void crashZone() {
    getAddress();
  }
```

Reply:
Oh ok, I see! Ok, keeping the check then.